qiskit/transpiler/passes/resource_estimation.py

```diff
@@ -0,0 +1,21 @@
+# -*- coding: utf-8 -*-
+
+# Copyright 2019, IBM.
+#
+# This source code is licensed under the Apache License, Version 2.0 found in
+# the LICENSE.txt file in the root directory of this source tree.
+
+""" Updates 'size', 'depth', and 'width' in the property set.
+"""
+from qiskit.transpiler.basepasses import AnalysisPass
+
+
+class ResourceEstimation(AnalysisPass):
+    """ Updates 'size', 'depth', and 'width' in the property set.
+    """
+
+    def run(self, dag):
+        self.property_set['size'] = dag.size()
+        self.property_set['depth'] = dag.depth()
+        self.property_set['width'] = dag.width()
```

Comment:
can you add these too:

`dag.num_tensor_factors()`  --> means number of separable subcircuits

`dag.count_ops()` --> returns a dictionary with a breakdown of how many gates of each kind (also addresses @nonhermitian's request).

Reply:
I was planing to remove `num_tensor_factors` https://github.com/Qiskit/qiskit-terra/issues/728#issuecomment-471074045

`count_ops` sees to be a pass, maybe it should be a different pass?

Reply:
`num_tensor_factors` is useful for understanding how entangled a circuit is (useful for the end user).

`count_ops` is basically resource estimation, it is giving a breakdown of the gates rather than an aggregate `size`. we should be careful to not end up with a million passes that are marginally different because it will make it harder to use.

Reply:
`count_ops` is literally a pass in the current `DAGCircuit`:
```
        op_dict = {}
        for node in self.node_nums_in_topological_order():
            nd = self.multi_graph.node[node]
            name = nd["name"]
            if nd["type"] == "op":
                if name not in op_dict:
                    op_dict[name] = 1
                else:
                    op_dict[name] += 1
        return op_dict
```

I will add `num_tensor_factors`.

Reply:
Yes but why can't this code be in the ResourceEstimation pass? It is literally doing resource estimation.

Getting depth is also internally a pass in networkx, we just don't see it as such because networkx provides a method to get it.

Reply:
I added `count_ops`. I did not added `num_tensor_factors`, since it might be a pass instead of a DAG method in the near future. What do you think?